Fix master ready status

Cynerva opened the pull request:
Fix for https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/143

1. Removed "Kubernetes master services ready." status, which occurred prematurely
2. Added "Waiting for kube-system pods to start" status

chuckbutler commented on cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py:
```diff
@@ -650,3 +643,24 @@ def setup_tokens(token, username, user):
         stream.write('{0},{1},{2}'.format(token, username, user))
+
+
+def all_kube_system_pods_running():
```
I love this! I think there's an opportunity for us to abstract and put this in a kubectl lib or we could potentially use the python kubernetes lib moving forward, but as a straw man, <3 <3 <3

chuckbutler commented on cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py:
```diff
@@ -166,7 +166,9 @@ def set_app_version():
-    if hookenv.config('service-cidr') != service_cidr():
+    if not all_kube_system_pods_running():
+        hookenv.status_set('waiting', 'Waiting for kube-system pods to start')
+    elif hookenv.config('service-cidr') != service_cidr():
```
I missed where we are declaring service_cidr(), this must have been a prior update that I missed where its now a method.

chuckbutler commented on cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py:
```diff
@@ -192,7 +194,7 @@ def start_master(etcd, tls):
         host.service_start(service)
     hookenv.open_port(6443)
-    hookenv.status_set('active', 'Kubernetes master services ready.')
+    hookenv.log('Kubernetes master services ready.')
```
but do we want this as a log? its not a problem but seems like its wasted data in the logging infra.

Cynerva replied:
I left it in as a log message to be cautious, but I think you're right, we don't really gain anything from it. I'll remove it!

chuckbutler commented on cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py:
```diff
@@ -309,8 +296,14 @@ def start_kube_dns():
-    create_addon('kubedns-svc.yaml', context)
+
+    try:
+        create_addon('kubedns-controller.yaml', context)
```
yes yes yes yes yes x 10000 <3 this code cleanup is good.

chuckbutler commented on cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py:
```diff
@@ -309,8 +296,14 @@ def start_kube_dns():
+        create_addon('kubedns-controller.yaml', context)
+        create_addon('kubedns-svc.yaml', context)
+    except CalledProcessError:
+        hookenv.log('kubedns deploy failed, will try again soon')
```
would it be prudent to notify via status that addon updates are failing atm? Open question, not a blocker. I like whats here.

Cynerva replied:
Good idea, though it's hard to give a good message that explains what we're waiting for. Usually `CalledProcessError` just means the API server isn't fully up yet, but it can mean other things too.

Probably just go with "Waiting to retry KubeDNS deployment" for now. I'll get that in there.

Cynerva commented:
Pushed additional changes:

1. Removed "Kubernetes master services ready." log message (PR feedback)
2. Changed "Rendering the Kubernetes DNS files." status to "Deploying KubeDNS"
3. Added "Waiting to retry KubeDNS deployment" status (PR feedback)
4. Wait for `kubernetes-master.components.started` state before we run `start_kube_dns`

wwwtyro commented:
LGTM

Cynerva commented:
Tested deployment with the latest changes. Here's the [juju status log](https://gist.github.com/Cynerva/e4b341d2dbe4bbc41331a1077d54ddc1) -- looks good to me.

Cynerva commented on cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py:
```diff
@@ -309,8 +293,14 @@ def start_kube_dns():
+        create_addon('kubedns-controller.yaml', context)
+        create_addon('kubedns-svc.yaml', context)
+    except CalledProcessError:
+        hookenv.status('waiting', 'Waiting to retry KubeDNS deployment')
```
Just spotted this mistake in the failure case, should be `status_set`. I'll fix that in the rebase against upstream.
